pkg/generators/forwarding/fluentd/generators_test.go

```diff
@@ -0,0 +1,166 @@
+package fluentd
+
+import (
+	"text/template"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	logging "github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1"
+	"github.com/openshift/cluster-logging-operator/pkg/generators"
+)
+
+var _ = Describe("Generating pipeline to output labels", func() {
+	var (
+		configGenerator *ConfigGenerator
+	)
+	BeforeEach(func() {
+		engn, err := generators.New("OutputLabelConf",
+			&template.FuncMap{
+				"labelName":           labelName,
+				"sourceTypelabelName": sourceTypeLabelName,
+			},
+			templateRegistry...)
+		Expect(err).To(BeNil())
+
+		configGenerator = &ConfigGenerator{
+			Generator:                  engn,
+			includeLegacyForwardConfig: false,
+			includeLegacySyslogConfig:  false,
+			useOldRemoteSyslogPlugin:   false,
+			storeTemplate:              "storeElasticsearch",
+			outputTemplate:             "outputLabelConf",
+		}
+	})
+
+	It("should generate no labels for a single pipeline", func() {
+		pipelines := []logging.PipelineSpec{
+			{
+				InputRefs:  []string{logging.InputNameInfrastructure},
+				OutputRefs: []string{"infra-es"},
+				Labels:     nil,
+				Name:       "Pipeline-1",
+			},
+		}
+		got, err := configGenerator.generatePipelineToOutputLabels(pipelines)
+		Expect(err).To(BeNil())
+		Expect(got[0]).To(BeEquivalentTo(`<label @PIPELINE_1>
+  <match **>
+    @type copy
+
+    <store>
+      @type relabel
+      @label @INFRA_ES
+    </store>
+  </match>
+</label>`))
+	})
+
+	It("should generate one label for a single pipeline", func() {
+		pipelines := []logging.PipelineSpec{
+			{
+				InputRefs:  []string{logging.InputNameInfrastructure},
+				OutputRefs: []string{"infra-es"},
+				Labels:     map[string]string{"1": "2"},
+				Name:       "Pipeline-1",
+			},
+		}
+		got, err := configGenerator.generatePipelineToOutputLabels(pipelines)
+		Expect(err).To(BeNil())
+		Expect(got[0]).To(BeEquivalentTo(`<label @PIPELINE_1>
+  <filter **>
+    @type record_transformer
+    <record>
+      openshift { "labels": {"1":"2"} }
+    </record>
+  </filter>
+  <match **>
+    @type copy
+
+    <store>
+      @type relabel
+      @label @INFRA_ES
+    </store>
+  </match>
+</label>`))
+	})
+
+	It("should generate multiple labels for a single pipeline", func() {
+		pipelines := []logging.PipelineSpec{
+			{
+				InputRefs:  []string{logging.InputNameInfrastructure},
+				OutputRefs: []string{"infra-es"},
+				Labels:     map[string]string{"1": "2", "3": "4"},
+				Name:       "Pipeline-1",
+			},
+		}
+		got, err := configGenerator.generatePipelineToOutputLabels(pipelines)
+		Expect(err).To(BeNil())
+		Expect(got[0]).To(BeEquivalentTo(`<label @PIPELINE_1>
```

Comment:
Consider using `Expect(results).To(EqualTrimLines(....))` from `"github.com/openshift/cluster-logging-operator/test"`. The diff is more concise and you can indent the expected literal string inside our code.

Reply:
The diff is more concise, I admit. But `EqualTrimLines` compares strings, not arrays of strings

Reply:
Ah good to know :+1:

Reply:
EqualTrimLines internally converts the multi-line string to a slice of lines and does the diff on the slice. Gingko matchers take their values as interface{}, so we could easily change EqualTrimLines to work for both strings and slices of strings.

Reply:
Fixed